`BETA` [debug] error: no declaration matches 'void esphome::debug::DebugComponent::on_shutdown()'

### The problem

I'm having error compiling with 2025.5.0b5 in some of my devices when using `debug` component:

```log
INFO ESPHome 2025.5.0b4
INFO Reading configuration /config/esphome/test-debug.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing test-debug (board: esp32dev; framework: arduino; platform: platformio/espressif32@5.4.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
Dependency Graph
|-- WiFi @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- Update @ 2.0.0
Compiling .pioenvs/test-debug/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/api_frame_helper.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/user_services.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/debug/debug_component.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/debug/debug_esp32.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/debug/debug_esp8266.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/debug/debug_host.cpp.o
src/esphome/components/debug/debug_esp32.cpp:50:6: error: no declaration matches 'void esphome::debug::DebugComponent::on_shutdown()'
 void DebugComponent::on_shutdown() {
      ^~~~~~~~~~~~~~
src/esphome/components/debug/debug_esp32.cpp:50:6: note: no functions named 'void esphome::debug::DebugComponent::on_shutdown()'
In file included from src/esphome/components/debug/debug_esp32.cpp:1:
src/esphome/components/debug/debug_component.h:18:7: note: 'class esphome::debug::DebugComponent' defined here
 class DebugComponent : public PollingComponent {
       ^~~~~~~~~~~~~~
Compiling .pioenvs/test-debug/src/esphome/components/debug/debug_libretiny.cpp.o
*** [.pioenvs/test-debug/src/esphome/components/debug/debug_esp32.cpp.o] Error 1
========================== [FAILED] Took 6.12 seconds ==========================
```

In other devices it works fine, but even creating a very simple new device I get this issue.

### Which version of ESPHome has the issue?

2025.5.0b4

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2025.5.2

### What platform are you using?

ESP32

### Board

esp32dev

### Component causing the issue

debug

### YAML Config

```yaml
esphome:
  name: test-debug
  friendly_name: test-debug

esp32:
  board: esp32dev

# Enable logging
logger:

# Enable Home Assistant API
api:

ota:
  - platform: esphome

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
 
debug:
```

### Anything in the logs that might be useful for us?

```txt
INFO ESPHome 2025.5.0b4
INFO Reading configuration /config/esphome/test-debug.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing test-debug (board: esp32dev; framework: arduino; platform: platformio/espressif32@5.4.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
Dependency Graph
|-- WiFi @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- Update @ 2.0.0
Compiling .pioenvs/test-debug/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/api_frame_helper.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/api/user_services.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/debug/debug_component.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/debug/debug_esp32.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/debug/debug_esp8266.cpp.o
Compiling .pioenvs/test-debug/src/esphome/components/debug/debug_host.cpp.o
src/esphome/components/debug/debug_esp32.cpp:50:6: error: no declaration matches 'void esphome::debug::DebugComponent::on_shutdown()'
 void DebugComponent::on_shutdown() {
      ^~~~~~~~~~~~~~
src/esphome/components/debug/debug_esp32.cpp:50:6: note: no functions named 'void esphome::debug::DebugComponent::on_shutdown()'
In file included from src/esphome/components/debug/debug_esp32.cpp:1:
src/esphome/components/debug/debug_component.h:18:7: note: 'class esphome::debug::DebugComponent' defined here
 class DebugComponent : public PollingComponent {
       ^~~~~~~~~~~~~~
Compiling .pioenvs/test-debug/src/esphome/components/debug/debug_libretiny.cpp.o
*** [.pioenvs/test-debug/src/esphome/components/debug/debug_esp32.cpp.o] Error 1
========================== [FAILED] Took 6.12 seconds ==========================
```

### Additional information

It works fine with ESPHome v2025.4.2.
It also fails with ESPHome v2025.6.0dev.

Could be relate to this?
- esphome/esphome#8595

--- a/esphome/components/debug/debug_component.h
+++ b/esphome/components/debug/debug_component.h
@@ -34,13 +34,15 @@ class DebugComponent : public PollingComponent {
 #endif
   void set_loop_time_sensor(sensor::Sensor *loop_time_sensor) { loop_time_sensor_ = loop_time_sensor; }
 #ifdef USE_ESP32
-  void on_shutdown() override;
   void set_psram_sensor(sensor::Sensor *psram_sensor) { this->psram_sensor_ = psram_sensor; }
 #endif  // USE_ESP32
   void set_cpu_frequency_sensor(sensor::Sensor *cpu_frequency_sensor) {
     this->cpu_frequency_sensor_ = cpu_frequency_sensor;
   }
 #endif  // USE_SENSOR
+#ifdef USE_ESP32
+  void on_shutdown() override;
+#endif  // USE_ESP32
  protected:
   uint32_t free_heap_{};
 

